syntax = "proto3";

import "google/protobuf/empty.proto";

package filesys;

service Filesys {
  rpc CreateFile(CreateFileArgs) returns (google.protobuf.Empty) {}

  rpc ReadBlocks(ReadBlocksArgs) returns (ReadBlocksReply) {}

  rpc WriteBlocks(WriteBlocksArgs) returns (google.protobuf.Empty) {}

  rpc GetFileList(GetFileListArgs) returns (GetFileListReply) {}

  rpc GetUpdateLog(GetUpdateLogArgs) returns (stream GetUpdateLogReply) {}

  rpc HeartBeat(google.protobuf.Empty) returns (stream HeartBeatReply) {}

  rpc DeleteFile(DeleteFileArgs) returns (google.protobuf.Empty) {}
}

message CreateFileArgs {
  string file_name = 1;
  bytes public_key = 2;
}

message ReadBlocksArgs {
  string file_name = 1;
  uint32 stripe_offset = 2;
  uint32 num_stripes = 3;
  optional uint32 version = 4;
}

message Metadata {
  bytes public_key = 1;
  uint64 file_size = 2;
}

message ReadBlocksReply {
  bytes block_data = 1;
  uint32 version = 2;
}

message WriteBlocksArgs {
  string file_name = 1;
  uint32 stripe_offset = 2;
  uint32 num_stripes = 3;
  uint32 version = 4;
  bytes block_data = 5;
  Metadata metadata = 6;
}

message GetFileListArgs {
  optional bool metadata = 1;
  optional string file_name = 2;
  optional bool include_delete = 3;
}

message FileInfo {
  string file_name = 1;
  uint32 version = 2;
  optional Metadata metadata = 3;
}

message GetFileListReply {
  repeated FileInfo files = 1;
}

message GetUpdateLogArgs {
  string file_name = 1;
  uint32 after_version = 2;
}

message GetUpdateLogReply {
  uint32 version = 1;
  uint32 stripe_offset = 2;
  uint32 num_stripes = 3;
}

message HeartBeatReply {
  string file_name = 1;
  uint32 version = 2;
}

message DeleteFileArgs {
  string file_name = 1;
}